### PR TITLE
API Overhaul

### PR DESCRIPTION
**Pull request by benfred**
First draft of the API overhauls changes. Adds most core functionality, including
defining workflow graphs with a ColumnGroup class, the workflow and dataset changes
, most operators converted to use the new api, etc.
- [x] ColumnGroup API
- [x] Workflow.fit/transform/fit_transform methods
- [x] Write Dataset to parquet/hugectr
- [x] Rename operator
- [x] Convert Core Operators to new api
- [x] Serialization (load/save statistics) (Ben)
- [x] Multi-column Categorify (combo/joint)
- [x] TargetEncoding op (Ben)
- [x] JoinGroupBy op
- [x] DropNa/Filter support (moved to https://github.com/NVIDIA/NVTabular/issues/515 )
- [x] LambdaOp dependencies (Ronay)
- [x] HashedCross w/ nested columns (Benedikt)
- [x] Feature Column conversion tool
- [x] Multihot embedding in TF/PyTorch Layer
- [x] Fix unittests
   - [x] test_column_similarity
   - [x] test_dask_nvt
   - [x] test_io
   - [x] test_notebooks (depends on notebooks)
   - [x] test_ops  (Benedikt)
   - [x] test_s3
   - [x] test_tf_dataloader (Ben)
   - [x] test_tf_layers
   - [x] test_torch_dataloader
   - [x] test_workflow  (Alberto)
- [x] Update Example notebooks
   - [x] Criteo (Benedikt)
   - [x] Rossmann (depends on serialization support) (Benedikt)
   - [x] Outbrain  (Ronay)
   - [x] Recsys2020. ( Benedikt )
   - [x] MovieLens (MultiHot) (Benedikt)
   - [x] Tensorflow notebooks (accelerating/feature column conversion) (Ronay)
- [x] get_embedding_sizes (Ben)
- [x] Handle hugectr cat/cont/labels metadata in parquet
- [x] Update documentation
   - [x] Update example docstrings in all operators (Benedikt)
   - [x] Verify README/ HowItWorks for references to old api (Ronay)
- [x] Benchmark performance relative to original version (Alberto)

### CODE REVIEW

**Comment by bschifferer** [comment]
I am working on the notebooks. There are some code broken with the PyTorch data loader:

image

**Comment by bschifferer** [comment]
I am not able to use the ops.DifferenceLag. However, I try to initalize it, I get an error

image
image
image

**Comment by bschifferer** [comment]
Converted many unittest to new API in https://github.com/NVIDIA/NVTabular/pull/496

**Comment by rnyak** [comment]
@benfred I can take Outbrain nb. What's with the  `Outbrain (depends on TargetEncoding op)` ? is it related to the error when we set kfold >1 ?

File: nvtabular/ops/target_encoding.py [hunk 1]
```diff
@@ -156,33 +139,21 @@ def __init__(
...
-                stat_name=self.stat_name,
-                name_sep=self.name_sep,
-                fold_name="__fold__",
-                kfold=self.kfold,
-                fold_seed=self.fold_seed,
-                fold_groups=self.cat_groups,
-            )
-        )
-        return stats
+    # TODO: fit/fit_finalize methods
```
Comment (rjzamora) [inline nvtabular/ops/target_encoding.py:142]: Just a note:  Converting `op_logic` to `transform` will have one complication with the new API...  We were previously storing a `"__fold__"` column during the "fit," and then using that column during the "transform."  We can/should no longer do that.  The new API is not designed to track an underlying ddf in the workflow (for good reason), and it was already bad for performnace in the old API.

Perhaps a reasonable workaround is to use a hashing of the local row index and some random seed: `f(<seed>, <local-row-index>) % <nfolds>`?

File: nvtabular/workflow.py [hunk 2]
```diff
@@ -13,1078 +13,178 @@
...
-from nvtabular.io.writer_factory import writer_factory
-from nvtabular.ops import DFOperator, Operator, StatOperator, TransformOperator
+from nvtabular.column_group import ColumnGroup, iter_nodes
+from nvtabular.io.dataset import Dataset
+from nvtabular.ops import StatOperator
 from nvtabular.worker import clean_worker_cache
 
-LOG = logging.getLogger("nvtabular")
+# import yaml
+
```
Comment (rjzamora) [inline nvtabular/workflow.py:28]:
```suggestion
```

File: nvtabular/workflow.py [hunk 3]
```diff
@@ -13,1078 +13,178 @@
...
-        fs = get_fs_token_paths(output_path)[0]
-        fs.mkdirs(output_path, exist_ok=True)
-        if shuffle or out_files_per_proc:
+    columns = list(flatten(cg.flattened_columns for cg in column_groups))
 
-            cat_names = self.get_final_cols_names("categorical")
-            cont_names = self.get_final_cols_names("continuous")
-            label_names = self.get_final_cols_names("label")
+    return ddf.map_partitions(
+        lambda gdf: _transform_partition(gdf, column_groups),
```
Comment (rjzamora) [inline nvtabular/workflow.py:157]:
```suggestion
        _transform_partition,
        column_groups,
```
I don't think this lambda is needed

**Review by rjzamora — COMMENTED** [review]
Sorry for taking so long to get to this.  I didn't have many useful comments yet, but I did look throught `workflow.py` (and some op/columngroup code).

I am very happy with the workflow code so far - Excellent work!  We will need to run some dask/criteo benchmarks to check for performance regressions, but I am not expecting any serious issues.  Overall, I am very excited about this :)

File: nvtabular/ops/target_encoding.py (hunk 1 above)
Comment (benfred) [reply nvtabular/ops/target_encoding.py]: Thanks for the suggestion! I've just converted the TargetEncoding op in the last commit there - and definitely had some issues with storing the ```__fold__``` operator.

I like the idea about hashing the local row index, but went the route of moving the '_add_fold' function out of the fit - and then sharing with transform.  I had to tweak it a bit to make _add_fold function deterministic - it was originally going 
```python
cupy.random.set_seed(seed)
cupy.random.choice((cupy.arange(kfold, dtype=typ), len(s))
```
which had a race condition when called in the fit function (w/ ddf.index.map_partitions). We had multiple threads going through that and setting the seed, which caused the unittest to fail

I fixed by changing to 
```python
state = cupy.random.RandomState(fold_seed)
return state.choice(cupy.arange(kfold, dtype=typ), len(s))
``` 
which fixed the unittests (not relying on a global seed which could change).

I'm *hoping* this is sufficient - all the TE unittests pass at least =)

**Comment by bschifferer** [comment]
@benfred JointGroupBy requires cont_names in the case of 'count' only. In the old API, we did not require to specify a cont column as `count` can be calculated with no continuous column.

image

**Comment by bschifferer** [comment]
A small Feature Request:
How do we keep columns in the output dataset, which are not part of the pipeline? I just want to pass them through.
Is it possible to allow that without explicit listing all columns?

**Comment by bschifferer** [comment]
JointGroupBy - output_path is not optional anymore

```
nvt.ops.JoinGroupby(cont_names=['reply'],stats=["count"], out_path='./')
```

File: nvtabular/ops/target_encoding.py (hunk 1 above)
Comment (rjzamora) [reply nvtabular/ops/target_encoding.py]: >I'm hoping this is sufficient - all the TE unittests pass at least =)

Cool - For this PR, we just want things to "work."  I don't feel that the original implementation was very well tuned anyway :)

**Comment by bschifferer** [comment]
@benfred 

There is a bug in the get_embedding_sizes function. The sorting is different to old API.

Old API: C1, C10, C11, C12, ..., C2, C21, C22, ..., C3, C4
New API: C1, C2, C3, C4, C5, C6

The issue is that the data loader/tabular model expect the embedding order based on the OLD API.

image

**Comment by benfred** [comment]
rerun tests

**Comment by benfred** [comment]
rerun tests

**Comment by benfred** [comment]
rerun tests

**Review by jperez999 — APPROVED** [review]
Changes I highlighted are small docs things and a couple of nitpicks.

File: examples/criteo-example.ipynb [hunk 4]
```diff
@@ -380,16 +563,55 @@
...
+     "execution_count": 20,
+     "metadata": {},
+     "output_type": "execute_result"
+    }
+   ],
    "source": [
-    "embeddings = list(get_embedding_sizes(proc).values())"
+    "embeddings = list(get_embedding_sizes(workflow).values())\n",
+    "# We limit the output dimension to 16\n",
+    "embeddings = [[emb[0], min(16, emb[1])] for emb in embeddings]\n",
```
[the remark this answers is not in the document]
Comment (benfred) [reply examples/criteo-example.ipynb:608]: I think using larger embedding sizes blows up the memory used during training - which is why we're making this change to be consistent with the old get_embedding_sizes function (which never returned embeddings bigger than 16 iirc). We should come back to this though - I think there is a better middle ground between the 16 we were returning and the 512 the new version will return.

File: HowItWorks.md [hunk 5]
```diff
@@ -1,79 +1,81 @@
...
+Operations are a reflection of the way in which compute happens on the GPU across large datasets. At a high level we’re concerned with two types of compute: the type that touches the entire dataset (or some large chunk of it) and the type that operates on a single row. Operations split the compute such that the first phase, which we call statistics gathering, is the only place where operations that cross the row boundary can take place. An example of this would be in the Normalize op which relies on two statistics, the mean and standard deviation. In order to normalize a row, we must first have calculated these two values, and we use a Dask-cudf graph to compute this part of the op.
```
[the remark this answers is not in the document]
Comment (benfred) [reply HowItWorks.md:12]: I'd like to rewrite this entire section  - lets leave in place for now,

File: HowItWorks.md [hunk 6]
```diff
@@ -1,79 +1,81 @@
...
-workflow.add_cont_preprocess(nvt.ops.Normalize())
+With the release of v0.4, we are extending our preprocessing and feature engineering workflows to be directed graphs of operators applied onto user defined groups of columns. Defining this graph is decoupled from the Workflow class, and let our users easily define complicated graphs of operations on their own custom defined sets of columns. Our Workflow is changed to adopt a similar API to that found with transformers in [scikit-learn](https://scikit-learn.org/stable/data_transforms.html): statistics will be calculated with a ‘fit’ method and applied with a ‘transform’ method. The NVTabular Dataset object is extended to handle both input and output of datasets with the ‘transform’ method of the workflow  taking an input Dataset and returning as output a transformed Dataset. 
 
-dataset = nvt.Dataset("/path/to/data.parquet")
+An op can be applied to a ColumnGroup from an overloaded >> operator, which in turn returns a new ColumnGroup, which further Operators can be applied on top of (see example below). A ColumnGroup is a list of string column names, and the operators work on every column in the ColumnGroup.
```
[the remark this answers is not in the document]
Comment (benfred) [reply HowItWorks.md:18]: thanks! fixed in next commit

File: examples/movielens_multihot_example.ipynb [hunk 7]
```diff
@@ -460,27 +460,28 @@
...
+   "cell_type": "markdown",
    "metadata": {},
-   "outputs": [],
    "source": [
-    "proc = nvt.Workflow(\n",
-    "    cat_names=['userId', 'movieId', 'genres'], \n",
-    "    cont_names=[], \n",
-    "    label_name=['rating']\n",
-    ")"
+    "Currently, our dataset are two separate dataframes. First, we use the `JoinExternal` operator to `left-join` the metadata (genres) to our rating dataset."
```
[the remark this answers is not in the document]
Comment (benfred) [reply examples/movielens_multihot_example.ipynb]: fixed in latest commit

File: nvtabular/ops/stat_operator.py [hunk 8]
```diff
@@ -18,33 +18,33 @@
...
-        )
+    def load(self, data):
+        """Loads statistics from a json-able blob of data. This is usually called
+        by the workflow rather than called directly"""
+        raise NotImplementedError(".load isn't implemented for this op!")
 
     def clear(self):
-        raise NotImplementedError("""zero and reinitialize all relevant statistical properties""")
+        """ zero and reinitialize all relevant statistical properties"""
+        raise NotImplementedError(".clear isn't implemented for this op!")
```
[the remark this answers is not in the document]
Comment (benfred) [reply nvtabular/ops/stat_operator.py]: fixed

File: examples/rossmann/rossmann-store-sales-fastai.ipynb [hunk 9]
```diff
@@ -418,178 +405,178 @@
        "  <tbody>\n",
        "    <tr>\n",
        "      <td>0</td>\n",
-       "      <td>0.802575</td>\n",
```
[the remark this answers is not in the document]
Comment (benfred) [reply examples/rossmann/rossmann-store-sales-fastai.ipynb:421]: good catch! fixed

File: examples/wnd_outbrain/Outbrain.ipynb [hunk 10]
```diff
@@ -731,28 +1012,7 @@
...
-      "Epoch 2/10\n",
-      "456/456 [==============================] - 164s 359ms/step - loss: 0.4272 - binary_accuracy: 0.8154 - auc: 0.7473 - val_loss: 0.4117 - val_binary_accuracy: 0.7902 - val_auc: 0.7332\n",
-      "Epoch 3/10\n",
-      "456/456 [==============================] - 164s 360ms/step - loss: 0.4214 - binary_accuracy: 0.8178 - auc: 0.7571 - val_loss: 0.4150 - val_binary_accuracy: 0.8159 - val_auc: 0.7493\n",
-      "Epoch 4/10\n",
-      "456/456 [==============================] - ETA: 0s - loss: 0.4157 - binary_accuracy: 0.8205 - auc: 0.7661"
-     ]
-    }
-   ],
+   "outputs": [],
```
[the remark this answers is not in the document]
Comment (benfred) [reply examples/wnd_outbrain/Outbrain.ipynb]: reran the notebook - has trained output now

File: examples/criteo-example.ipynb [hunk 11]
```diff
@@ -399,15 +621,85 @@
...
+       "</table>"
+      ],
+      "text/plain": [
+       "<IPython.core.display.HTML object>"
+      ]
+     },
+     "metadata": {},
+     "output_type": "display_data"
+    },
+    {
```
[the remark this answers is not in the document]
Comment (benfred) [reply examples/criteo-example.ipynb]: Ran this notebook again - no longer has an exception in there now (though I didn't wait for it to finish so training isn't complete)
